### PR TITLE
Add support contact email

### PR DESCRIPTION
**Pull request by arothuis**
[Ticket 110497290](https://www.pivotaltracker.com/story/show/110497290) and [106203816](https://www.pivotaltracker.com/story/show/106203816)

### CODE REVIEW

**Comment by arothuis** [comment]
Will check it myself after Christmas :santa:, will ping when ready for review.

**Comment by thijskh** [comment]
So I understand the engineblock entityid is needed because in theory it could be in the list of authenticating authorities in the assertions you receive? Because in practice, it is not:

```xml
            <saml:AuthnContext>
                <saml:AuthnContextClassRef>urn:oasis:names:tc:SAML:2.0:ac:classes:Password</saml:AuthnContextClassRef>
                <saml:AuthenticatingAuthority>https://idp.surfnet.nl</saml:AuthenticatingAuthority>
            </saml:AuthnContext>
```

File: src/OpenConext/EngineBlockApiClientBundle/Repository/ContactPersonRepository.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+    /**
+     * @param EntityId $entityId
+     * @return ContactPersonList
+     */
+    public function findAll(EntityId $entityId)
+    {
+        $identityProviderJson = $this->apiClient->read('metadata/idp?entity-id=%s', [$entityId->getEntityId()]);
+
+        return ContactPersonListFactory::create($identityProviderJson);
```
Comment (rjkip) [inline src/OpenConext/EngineBlockApiClientBundle/Repository/ContactPersonRepository.php]: I don't like the name `create`. We're not creating a contact person list factory, and the input data is specific: deserialised JSON output from the IdP metadata API. Maybe `createListFromMetadata`?

**Comment by arothuis** [comment]
@thijskh, you understand correctly. In order to get the underlying authenticating authority, we need to be sure EngineBlock is not part of the list of authenticating authorities. 

A small check has been implemented to verify and ensure this, so that even when an altered EngineBlock is used that does add itself as an authenticating authority our assumptions still stand.

File: src/OpenConext/EngineBlockApiClientBundle/Value/ContactPersonListFactory.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    public static function create(array $data)
+    {
+        Assert::isArray($data, 'Contact list JSON structure must be an associative array, got %s');
+        Assert::keyExists($data, 'contact_persons', 'Entity JSON structure must contain key "contact_persons"');
+
+        // We cannot use self::class because translation extractions fails on that
+        $contactPersons = array_map(
+            ['\OpenConext\EngineBlockApiClientBundle\Value\ContactPersonListFactory', 'createContactPerson'],
+            $data['contact_persons']
+        );
```
Comment (rjkip) [inline src/OpenConext/EngineBlockApiClientBundle/Value/ContactPersonListFactory.php]: This mapping can fail with a generic PHP error message when one of the contact persons is not an array. I'd rather have an explicit assert with a relevant error message.

File: src/OpenConext/ProfileBundle/Service/SupportContactEmailService.php [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        $contactPersons = $this->contactPersonRepository->findAll($entityId);
+
+        if (count($contactPersons) === 0) {
+            return null;
+        }
+
+        $supportContactPersons = $contactPersons->filter(
+            function (ContactPerson $contactPerson) {
+                return $contactPerson->hasContactTypeOf(new ContactType(ContactType::TYPE_SUPPORT))
+                && $contactPerson->hasEmailAddress();
```
Comment (rjkip) [inline src/OpenConext/ProfileBundle/Service/SupportContactEmailService.php]: :rage1:

File: src/OpenConext/ProfileBundle/Service/SupportContactEmailService.php [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+     */
+    public function findSupportContactEmail(EntityId $entityId)
+    {
+        $contactPersons = $this->contactPersonRepository->findAll($entityId);
+
+        if (count($contactPersons) === 0) {
+            return null;
+        }
+
+        $supportContactPersons = $contactPersons->filter(
```
Comment (rjkip) [inline src/OpenConext/ProfileBundle/Service/SupportContactEmailService.php]: This filter can be appended to the `findAll` above :wink:

File: src/OpenConext/Profile/Value/ContactPersonList.php [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+                    return $predicate($contactPerson);
+                }
+            )
+        );
+    }
+
+    /**
+     * @return ContactPerson
+     */
+    public function getFirstContactPerson()
```
Comment (rjkip) [inline src/OpenConext/Profile/Value/ContactPersonList.php]: It's quite common to name this `first()` (see Doctrine Common). The `ContactPerson` parts seems kind of verbose.

File: src/OpenConext/Profile/Value/ContactType.php [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    public function __construct($contactType)
+    {
+        Assert::string($contactType);
+        Assert::choice($contactType, [
+            self::TYPE_TECHNICAL,
+            self::TYPE_SUPPORT,
+            self::TYPE_ADMINISTRATIVE,
+            self::TYPE_BILLING,
+            self::TYPE_OTHER,
+        ], '"%s" is not one of the valid ContactTypes: %s');
```
Comment (rjkip) [inline src/OpenConext/Profile/Value/ContactType.php]: Nit: not PSR-2 ;)

File: src/OpenConext/Profile/Value/ContactType.php [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    const TYPE_ADMINISTRATIVE = 'administrative';
+    const TYPE_BILLING = 'billing';
+    const TYPE_OTHER = 'other';
+
+    /**
+     * @var string
+     */
+    private $contactType;
+
+    public function __construct($contactType)
```
Comment (rjkip) [inline src/OpenConext/Profile/Value/ContactType.php]: Nit: doc

File: src/OpenConext/Profile/Value/ContactPersonList.php (hunk 5 above)
Comment (rjkip) [inline src/OpenConext/Profile/Value/ContactPersonList.php]: Also, what happens when there's no first contact person?

**Comment by arothuis** [comment]
@rjkip

File: src/OpenConext/Profile/Value/ContactPersonList.php [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace OpenConext\Profile\Value;
+
+use ArrayIterator;
+use Countable;
+use IteratorAggregate;
+use OpenConext\Profile\Exception\OutOfBoundsException;
```
Comment (rjkip) [inline src/OpenConext/Profile/Value/ContactPersonList.php]: This should be removed. After that, :+1:

**Pull request merged**